rfcs/2020-10-06-3791-composing-components-pt-1.md

```diff
@@ -0,0 +1,87 @@
+# RFC 3791 - 2020-10-06 - Composing Components: Part 1
+
+Vector is designed to be very modular, and the current tool for composing those
+modules is the TOML config file. This gives users a great deal of flexibility,
+but it can require configurations that are a bit verbose and require more of
+users than other pre-built, specific solutions.
+
+One way that Vector could get some of the best of both worlds would be to make
+it easy to create pre-built "chunks" of config that users could configure as
+normal components. These would be bundles of lower-level components wired
+together with adjusted default values for the specific use case.
+
+## Motivation
+
+We need a way to rapidly assemble Vector components that address specific use
+cases. This will allow us to improve ease of use without spending significant
+development time on each individual use case. It will allow us to focus
+development time on reuseable components without forcing users to do the work of
+assembling them from scratch.
+
+## Internal Proposal
+
+There are multiple levels at which we could implement this type of
+functionality:
+
+1. Manually implement new component as config facade over one existing component
+2. Manually implement new component as config facade over one source and one
+   parsing transform
+3. Manually implement new component as config expanding to arbitrary pipeline of
+   components
+4. Automatically derive new component from data describing arbitrary pipeline of
+   components
+
+We currently are at level (1), where we can do things like implement the Humio
+sink as a wrapper around the existing Splunk HEC sink.
+
+The next simplest is level (2). While it's not implemented yet, we do have
+existing plans to introduce the idea of a codec attached to sources. This would
+allow users to directly configure how to parse the incoming data as part of the
+source config itself. With that feature implemented, it would be relatively
+straightforward to do something similar to level (1) but expanding to both
+a source and an included codec.
```

Comment:
We should also consider sinks when discussing codecs. :) Seems like that work is left for a pt 2 though?

Reply:
Yes, codecs are basically a way for users to do something similar manually. Definitely related, but achieves a different goal.